qa/tasks/cephfs/test_client_recovery.py

```diff
@@ -755,3 +756,104 @@ def test_refuse_client_session_on_reconnect(self):
+        """
+        If any client gets unresponsive/it's session get idle due to lagginess
+        with any OSD and if config option defer_client_eviction_on_laggy_osds
+        is set true(default true) then make sure clients are not evicted until
+        OSD(s) return to normal.
+        """
+
+        self.fs.mds_asok(['config', 'set', 'mds_defer_session_stale', 'false'])
+        self.assertEqual(self.config_get(
+            'mds', 'defer_client_eviction_on_laggy_osds'), 'true')
+
+        # make an OSD laggy
+        osd_processes = self.mount_a.run_shell(
+            "pgrep -a ceph-osd",
+            stdout=StringIO()).stdout.getvalue().strip().splitlines()
+        self.assertNotEqual(len(osd_processes), 0, "No OSD process found")
+        osd = osd_processes[0].split(' ')[0]
+        halt_osd = self.mount_a.run_shell(f"sudo kill -SIGSTOP {osd}").returncode
+        self.assertEqual(halt_osd, 0, f"error while halting osd process {osd}")
+        time.sleep(120)
+        start_osd = self.mount_a.run_shell(f"sudo kill -SIGCONT {osd}").returncode
+        self.assertEqual(start_osd, 0, f"error while starting osd process {osd}")
+
+        try:
+            mount_b_gid = self.mount_b.get_global_id()
+
+            self.mount_b.kill()
+
+            # client session should be open, it gets stale
+            # only after session_timeout time.
+            self.assert_session_state(mount_b_gid, "open")
+
+            # makes session stale
+            time.sleep(self.fs.get_var("session_timeout") * 1.5)
+            self.assert_session_state(mount_b_gid, "stale")
+
+            # it takes time to have laggy clients entries in cluster log,
+            # wait for 6 minutes to see if it is visible, finally restart
+            # the client
+            tries = 6
+            while True:
+                try:
+                    with self.assert_cluster_log("1 client(s) laggy due to laggy OSDs",
+                                                 timeout=55):
+                        # make sure clients weren't evicted
+                        self.assert_session_count(2)
+                        break
+                except AssertionError:
+                    tries -= 1
+                    if tries:
+                        continue
+                    raise
+        finally:
+            self.mount_b.kill_cleanup()
+            self.mount_b.mount_wait()
+            self.mount_b.create_destroy()
+
+    def test_client_on_laggy_osd_2(self):
+        """
+        Since cluster is not reset, OSD that was made laggy in test_1 would
+        still be laggy(its laggy values would still be non-zero) thus this
+        test case needs to run after it and make sure that even if an OSD
+        is laggy but config option defer_client_eviction_on_laggy_osds is
+        set false then an unresponsive client does get evicted.
+        """
+
+        self.fs.mds_asok(['config', 'set', 'mds_defer_session_stale', 'false'])
```

Comment:
Spoke to @dparmar18 verbally -- this test is written with an assumption that it will be executed after `test_client_on_laggy_osd_1` (assumption being that the laggy OSD state would be available in this test too) - which is not a great way to writes test. Test cases need to be independent. You'd need to ensure that the tests leaves the cluster state as it was when the test starts execution (which for most test case would be a nice a cluster cluster).

Reply:
I agree

Reply:
@dparmar mentioned that there is no easy way to mark an OSD an un-laggy and @rzarzynski recommended to introduce a CLI to support this. I have a question - if an OSD becomes laggy for reasons such as failed network and then when the connection restores, does the OSD remain laggy and requires a manual intervention? I guess that once the OSD catches up on IOs from its peers in its PGs, it should automatically get marked as un-laggy. Isn't that the case @rzarzynski?

Reply:
@gregsfortytwo ^^

Reply:
Yeah. OSDs will lose their laggy state automatically, though I don't know the exact conditions.